src/Microsoft.Diagnostics.DebugServices.Implementation/ServiceManager.cs

```diff
@@ -232,30 +311,79 @@ public void LoadExtension(string extensionPath)
+        /// when running on desktop Framework.
         /// </summary>
-        /// <param name="assembly">extension assembly</param>
-        public void RegisterAssembly(Assembly assembly)
+        /// <param name="extensionPath">extension assembly path</param>
+        /// <returns>assembly</returns>
+        private Assembly UseAssemblyLoadContext(string extensionPath)
         {
-            if (_finalized)
+            ExtensionLoadContext extension = new(extensionPath);
+            Assembly assembly = extension.LoadFromAssemblyPath(extensionPath);
+            if (assembly is not null)
             {
-                throw new InvalidOperationException();
+                // This list is just to keep the load context alive
+                _extensions.Add(extension);
             }
+            return assembly;
+        }
 
-            try
+        private sealed class ExtensionLoadContext : AssemblyLoadContext
+        {
+            private static readonly HashSet<string> s_defaultAssemblies = new() {
+                "Microsoft.Diagnostics.DebugServices",
+                "Microsoft.Diagnostics.DebugServices.Implementation",
+                "Microsoft.Diagnostics.ExtensionCommands",
+                "Microsoft.Diagnostics.NETCore.Client",
+                "Microsoft.Diagnostics.Repl",
+                "Microsoft.Diagnostics.Runtime",
+                "Microsoft.FileFormats",
+                "Microsoft.SymbolStore",
+                "SOS.Extensions",
+                "SOS.Hosting",
+                "SOS.InstallHelper"
+            };
+
+            private Dictionary<string, string> _extensionPaths;
+
+            public string ExtensionPath { get; }
+
+            public ExtensionLoadContext(string extensionPath)
             {
-                RegisterExportedServices(assembly);
-                _notifyExtensionLoad.Fire(assembly);
+                ExtensionPath = extensionPath;
             }
-            catch (Exception ex) when (ex is DiagnosticsException or NotSupportedException or FileNotFoundException)
+
+            protected override Assembly Load(AssemblyName assemblyName)
             {
-                Trace.TraceError(ex.ToString());
+                lock (this)
+                {
+                    if (_extensionPaths == null)
+                    {
+                        string[] extensionFiles = Directory.GetFiles(Path.GetDirectoryName(ExtensionPath), "*.dll");
+                        _extensionPaths = new Dictionary<string, string>();
+                        foreach (string file in extensionFiles)
+                        {
+                            _extensionPaths.Add(Path.GetFileNameWithoutExtension(file), file);
+                        }
+                    }
+                }
+                if (s_defaultAssemblies.Contains(assemblyName.Name))
+                {
+                    string assemblyPath = Path.GetDirectoryName(Assembly.GetExecutingAssembly().Location);
```

Comment:
We could probably cache this in a static. This check also likely won't work on VS. Does it work on WinDBG?